File: dmn-test-cases/standard/tck/1.3/cl3/0085-decision-services/translator/expected/python/dmn/DecisionService_015.py
import typing
import decimal
import datetime
import isodate
import time

import jdmn.runtime.Context
import jdmn.runtime.DefaultDMNBaseDecision
import jdmn.runtime.ExecutionContext
import jdmn.runtime.LambdaExpression
import jdmn.runtime.LazyEval
import jdmn.runtime.Pair
import jdmn.runtime.Range
import jdmn.runtime.RuleOutput
import jdmn.runtime.RuleOutputList

import jdmn.runtime.annotation.Annotation
import jdmn.runtime.annotation.AnnotationSet
import jdmn.runtime.annotation.DRGElementKind
import jdmn.runtime.annotation.ExpressionKind
import jdmn.runtime.annotation.HitPolicy

import jdmn.runtime.cache.Cache

import jdmn.runtime.external.ExternalFunctionExecutor

import jdmn.runtime.listener.Arguments
import jdmn.runtime.listener.DRGElement
import jdmn.runtime.listener.EventListener
import jdmn.runtime.listener.Rule

import Decision_015_1
import Decision_015_2


# Generated(value = ["ds.ftl", "decisionService_015"])
class DecisionService_015(jdmn.runtime.DefaultDMNBaseDecision.DefaultDMNBaseDecision):
    DRG_ELEMENT_METADATA: jdmn.runtime.listener.DRGElement.DRGElement = jdmn.runtime.listener.DRGElement.DRGElement(
        "http://www.montera.com.au/spec/DMN/0085-decision-services",
        "decisionService_015",
        "",
        jdmn.runtime.annotation.DRGElementKind.DRGElementKind.DECISION_SERVICE,
        jdmn.runtime.annotation.ExpressionKind.ExpressionKind.OTHER,
        jdmn.runtime.annotation.HitPolicy.HitPolicy.UNKNOWN,
        -1
    )
    _instance = None

    def __init__(self):
        raise RuntimeError("Call instance() instead")

    @classmethod
    def instance(cls):
        if cls._instance is None:
            cls._instance = cls.__new__(cls)
            jdmn.runtime.DefaultDMNBaseDecision.DefaultDMNBaseDecision.__init__(cls._instance)
            cls._instance.initSubDecisions()
        return cls._instance

    def initSubDecisions(self, decision_015_1: Decision_015_1.Decision_015_1 = None, decision_015_2: Decision_015_2.Decision_015_2 = None):
        self.decision_015_1 = Decision_015_1.Decision_015_1() if decision_015_1 is None else decision_015_1
        self.decision_015_2 = Decision_015_2.Decision_015_2() if decision_015_2 is None else decision_015_2

    def apply(self, context_: jdmn.runtime.ExecutionContext.ExecutionContext) -> typing.Optional[jdmn.runtime.Context.Context]:
        try:
            # Start DS 'decisionService_015'
            annotationSet_: jdmn.runtime.annotation.AnnotationSet.AnnotationSet = None if context_ is None else context_.annotations
            eventListener_: jdmn.runtime.listener.EventListener.EventListener = None if context_ is None else context_.eventListener
            externalExecutor_: jdmn.runtime.external.ExternalFunctionExecutor.ExternalFunctionExecutor = None if context_ is None else context_.externalFunctionExecutor
            cache_: jdmn.runtime.cache.Cache.Cache = None if context_ is None else context_.cache
            decisionService_015StartTime_ = int(time.time_ns()/1000)
            decisionService_015Arguments_ = jdmn.runtime.listener.Arguments.Arguments()
            eventListener_.startDRGElement(self.DRG_ELEMENT_METADATA, decisionService_015Arguments_)

            # Evaluate DS 'decisionService_015'
            output_: typing.Optional[jdmn.runtime.Context.Context] = self.evaluate(context_)

            # End DS 'decisionService_015'
            eventListener_.endDRGElement(self.DRG_ELEMENT_METADATA, decisionService_015Arguments_, output_, (int(time.time_ns()/1000) - decisionService_015StartTime_))

            return output_
        except Exception as e:
            self.logError("Exception caught in 'decisionService_015' evaluation", e)
            return None

    def evaluate(self, context_: jdmn.runtime.ExecutionContext.ExecutionContext) -> typing.Optional[jdmn.runtime.Context.Context]:
        # Apply child decisions
        decision_015_1: typing.Optional[str] = self.decision_015_1.apply(context_)
        decision_015_2: typing.Optional[str] = self.decision_015_2.apply(context_)

        output_: typing.Optional[jdmn.runtime.Context.Context] = jdmn.runtime.Context.Context()
        output_.add("decision_015_1",  decision_015_1)
        output_.add("decision_015_2",  decision_015_2)
        return output_

File: dmn-test-cases/standard/tck/1.3/cl3/0085-decision-services/translator/expected/java/dmn/DecisionService_015.java

import java.util.*;
import java.util.stream.Collectors;

@javax.annotation.Generated(value = {"ds.ftl", "decisionService_015"})
@com.gs.dmn.runtime.annotation.DRGElement(
    namespace = "http://www.montera.com.au/spec/DMN/0085-decision-services",
    name = "decisionService_015",
    label = "",
    elementKind = com.gs.dmn.runtime.annotation.DRGElementKind.DECISION_SERVICE,
    expressionKind = com.gs.dmn.runtime.annotation.ExpressionKind.OTHER,
    hitPolicy = com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
    rulesCount = -1
)
public class DecisionService_015 extends com.gs.dmn.runtime.JavaTimeDMNBaseDecision<com.gs.dmn.runtime.Context> {
    public static final com.gs.dmn.runtime.listener.DRGElement DRG_ELEMENT_METADATA = new com.gs.dmn.runtime.listener.DRGElement(
        "http://www.montera.com.au/spec/DMN/0085-decision-services",
        "decisionService_015",
        "",
        com.gs.dmn.runtime.annotation.DRGElementKind.DECISION_SERVICE,
        com.gs.dmn.runtime.annotation.ExpressionKind.OTHER,
        com.gs.dmn.runtime.annotation.HitPolicy.UNKNOWN,
        -1
    );

    private static class DecisionService_015LazyHolder {
        static final DecisionService_015 INSTANCE = new DecisionService_015();
    }
    public static DecisionService_015 instance() {
        return DecisionService_015LazyHolder.INSTANCE;
    }

    private final Decision_015_1 decision_015_1;
    private final Decision_015_2 decision_015_2;

    private DecisionService_015() {
        this(new Decision_015_1(), new Decision_015_2());
    }

    private DecisionService_015(Decision_015_1 decision_015_1, Decision_015_2 decision_015_2) {
        this.decision_015_1 = decision_015_1;
        this.decision_015_2 = decision_015_2;
    }

    @java.lang.Override()
    public com.gs.dmn.runtime.Context applyMap(java.util.Map<String, String> input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return apply(context_);
        } catch (Exception e) {
            logError("Cannot apply decision 'DecisionService_015'", e);
            return null;
        }
    }

    @java.lang.Override()
    public com.gs.dmn.runtime.Context applyPojo(com.gs.dmn.runtime.ExecutableDRGElementInput input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return apply(context_);
        } catch (Exception e) {
            logError("Cannot apply element 'DecisionService_015'", e);
            return null;
        }
    }

    @java.lang.Override()
    public com.gs.dmn.runtime.Context applyContext(com.gs.dmn.runtime.Context input_, com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            return applyPojo(new DecisionService_015Input_(input_), context_);
        } catch (Exception e) {
            logError("Cannot apply element 'DecisionService_015'", e);
            return null;
        }
    }

    public com.gs.dmn.runtime.Context apply(com.gs.dmn.runtime.ExecutionContext context_) {
        try {
            // Start DS 'decisionService_015'
            com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
            com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
            com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
            com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;
            long decisionService_015StartTime_ = System.currentTimeMillis();
            com.gs.dmn.runtime.listener.Arguments decisionService_015Arguments_ = new com.gs.dmn.runtime.listener.Arguments();
            eventListener_.startDRGElement(DRG_ELEMENT_METADATA, decisionService_015Arguments_);

            // Evaluate DS 'decisionService_015'
            com.gs.dmn.runtime.Context output_ = lambda.apply(context_);

            // End DS 'decisionService_015'
            eventListener_.endDRGElement(DRG_ELEMENT_METADATA, decisionService_015Arguments_, output_, (System.currentTimeMillis() - decisionService_015StartTime_));

            return output_;
        } catch (Exception e) {
            logError("Exception caught in 'decisionService_015' evaluation", e);
            return null;
        }
    }

    public com.gs.dmn.runtime.LambdaExpression<com.gs.dmn.runtime.Context> lambda =
        new com.gs.dmn.runtime.LambdaExpression<com.gs.dmn.runtime.Context>() {
            public com.gs.dmn.runtime.Context apply(Object... args_) {
                com.gs.dmn.runtime.ExecutionContext context_ = 0 < args_.length ? (com.gs.dmn.runtime.ExecutionContext) args_[0] : null;
                com.gs.dmn.runtime.annotation.AnnotationSet annotationSet_ = context_ != null ? context_.getAnnotations() : null;
                com.gs.dmn.runtime.listener.EventListener eventListener_ = context_ != null ? context_.getEventListener() : null;
                com.gs.dmn.runtime.external.ExternalFunctionExecutor externalExecutor_ = context_ != null ? context_.getExternalFunctionExecutor() : null;
                com.gs.dmn.runtime.cache.Cache cache_ = context_ != null ? context_.getCache() : null;

                // Apply child decisions
                String decision_015_1 = DecisionService_015.this.decision_015_1.apply(context_);
                String decision_015_2 = DecisionService_015.this.decision_015_2.apply(context_);

                com.gs.dmn.runtime.Context output_ = new com.gs.dmn.runtime.Context();
                output_.add("decision_015_1",  decision_015_1);
                output_.add("decision_015_2",  decision_015_2);
                return output_;
            }
    };
}